### PR TITLE
TVM SGX

### LINKED ISSUE #912: [RUNTIME][SECURITY] Isolate IO dependent code
This is issue bought up by @nhynes. It is good practice to isolate the IO dependent code from the minimum runtime system, to provide a version that does not touch IO at all, this is useful for cases like emscripten and secured computing.

- [ ] Move the GetFileFormat out from file_util.h as it is not IO dependent
- [ ] upstream in dmlc, conditionally compile some of the checks

### PR DESCRIPTION
**Pull request by nhynes**
This PR allows compiling a TVM runtime that can be used in an Intel SGX enclave. It depends on https://github.com/dmlc/dmlc-core/pull/368.

Re #912: using guards seemed easier than splitting the header files (which would make understanding the codebase harder for new users, I think). I'm open to suggestions, though!

Still TODO: add working example in `app/sgx`

Resolves #912

### CODE REVIEW

**Comment by nhynes** [comment]
@tqchen The example in `apps/sgx` should work now. Please let me know what you think. If it's to your liking, I'll tidy things up and add some docs!

**Comment by tqchen** [comment]
In general, it looks good. To make the code style in apps/sgx also consistent with Google C style used in the rest of the project, let us change the suffix from cpp to cc, unless there is a code style culture for SGX that suggest the other way.

**Comment by nhynes** [comment]
re: the failing tests, any tips? they work fine on my machine, so it's tough to figure out how (whether?) to fix them.

Otherwise, I'll have the PR spiffied up and ready to go by tomorrow.

**Comment by nhynes** [comment]
Okay, the code is now ready for a proper review! At your leisure.

File: include/tvm/runtime/packed_func.h [hunk 1]
```diff
@@ -674,10 +675,12 @@ inline TVMType String2TVMType(std::string s) {
     scan = s.c_str();
     LOG(FATAL) << "unknown type " << s;
   }
-  unsigned bits = t.bits, lanes = t.lanes;
-  sscanf(scan, "%ux%u", &bits, &lanes);
-  t.bits = static_cast<uint8_t>(bits);
-  t.lanes = static_cast<uint16_t>(lanes);
+  unsigned bits, lanes;
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h]: This code need to be change to support all possible cases

- ```int```  (no bits and lanes)
- ```int32``` (only bits, no lanes)
- ```int32x4``` (both bits and lanes)

so far it only support the last version

Comment (tqchen) [reply include/tvm/runtime/packed_func.h]: This can be the main reason causing the CI to fail

File: include/tvm/runtime/packed_func.h [hunk 2]
```diff
@@ -640,20 +640,21 @@ inline const char* TypeCode2Str(int type_code) {
...
+#ifndef _LIBCPP_SGX_NO_IOSTREAMS
 inline std::ostream& operator<<(std::ostream& os, TVMType t) {  // NOLINT(*)
-  os << TypeCode2Str(t.code);
-  if (t.code == kHandle) return os;
-  os << static_cast<int>(t.bits);
-  if (t.lanes != 1) {
-    os << 'x' << static_cast<int>(t.lanes);
-  }
-  return os;
+  return os << TVMType2String(t);
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h]: Keep the original content of the serialization, even if it is a bit duplicating, this one can be more efficient

File: include/tvm/runtime/packed_func.h [hunk 3]
```diff
@@ -640,20 +640,21 @@ inline const char* TypeCode2Str(int type_code) {
...
 }
+#endif
 
 inline std::string TVMType2String(TVMType t) {
-  std::ostringstream os;
-  os << t;
-  return os.str();
+  std::string repr = "";
+  repr += TypeCode2Str(t.code);
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h]: Use this version for SGX variant

File: include/tvm/runtime/packed_func.h (hunk 2 above)
Comment (tqchen) [reply include/tvm/runtime/packed_func.h]: keep original variant of TVMType2String here

File: include/tvm/runtime/packed_func.h [hunk 4]
```diff
@@ -675,12 +687,20 @@ inline TVMType String2TVMType(std::string s) {
...
+  sscanf(scan, "%ux%u", &bits, &lanes);
+  t.bits = static_cast<uint8_t>(bits);
+  t.lanes = static_cast<uint16_t>(lanes);
+#else
+  char* xdelim;
   bits = strtoul(scan, &xdelim, 10);
-  lanes = strtoul(xdelim + 1, nullptr, 10);
   if (bits != 0) t.bits = static_cast<uint8_t>(bits);
-  if (lanes != 0) t.lanes = static_cast<uint8_t>(lanes);
+  if (scan[xdelim] == 'x') {
```
Comment (tqchen) [inline include/tvm/runtime/packed_func.h]: xdelim doesn't return the location of the index, but returns the pointer. We should do instead

if (*xdelim == 'x') {
}

Comment (tqchen) [reply include/tvm/runtime/packed_func.h]: Let us still use your logic to replace the sscanf one, as this one seems to be as efficient and we can test it via CI

**Pull request merged**